### DISCUSSION TITLE
Is it possible to protect static JS chunks?

### DISCUSSION THREAD

**Discussion by ghost** [vercel/next.js#33114] [Help]
Hey everyone,

We are in the process of migrating our React app to Next.js.
Currently, we have a BFF protecting our app and assets with auth0.
We do use Auth0's lib to protect our app, but unfortunately, it does not protect the js chunks, so getting them is as easy as navigating to `https://my-app.com/_next/static/chunks/pages/%5B%5B...app%5D%5Da0a00sd0.js`

Is there any way I will be able to protect it without a custom server?

Thanks in advance!

**Comment by JustFly1984** [comment]
You can't hide the code, it is against whole scripting language nature.
The thing you actually want is to secure the code from injection and unauthorised changes. Please look into CSP headers and nonce signature.

**Comment by ghost** [comment]
@JustFly1984  Wdym? We actually did it.
We had a server that validated the authentication before serving anything from `/`.

How can CSP headers help here?

**Comment by ghost** [comment]
Bumping ❤️

**Reply by U-4-E-A** [reply]
Could I ask for more detail as to how you implemented your solution? I would like to protect these assets on a basis of both authentication and authorisation, the latter I think maybe be difficult or impossible. I am thinking about using AWS S3 signed cookies/signed URL as an option but I think this would only work for authentication.

**Reply by ghost** [reply]
@U-4-E-A hmm we did not deploy our solution yet (ye, since), but I eventually created a middleware under `_next` folder where I implemented the asset protection.

**Reply by U-4-E-A** [reply]
> @U-4-E-A hmm we did not deploy our solution yet (ye, since), but I eventually created a middleware under `_next` folder where I implemented the asset protection.

Can you expand on how you did this? I am still looking into a Cloudfront solution. Are you using next server as your only server or are you using a backend server as well? Thanks.

**Reply by U-4-E-A** [reply]
One thing I am thinking about doing (if I have this right) is setting my Cloudfront alias to that of my domain, allowing the passing of cookies (for the session cookie) to Cloudfront. I can then make a request to my backend server (using graphql), which will pass the session cookie. I can then confirm both the authentication of the user and authorisation of access by looking up the user's role in database.

**Comment by demircancelebi** [comment]
I was trying to secure a specific file like `_next/static/chunks/pages/my-private-page-somehash.js`

I solved it by creating an `api/static` endpoint, and adding a redirect to `.next.config.mjs`

```ts
// src/pages/api/static.ts

import fs from "fs";
import path from "path";
import { type NextApiRequest, type NextApiResponse } from "next";
import { getToken } from "next-auth/jwt";

const handler = async (req: NextApiRequest, res: NextApiResponse) => {
  const { path: thePath } = req.query;

  const token = await getToken({ req, secret: process.env.SECRET });

  if ((thePath as string).startsWith("my-private-page-") && !token) {
    res.status(401).json({ error: "Unauthorized" });
    return;
  }

  const file = fs.readFileSync(
    path.join(process.cwd(), ".next/static/chunks/pages", thePath as string)
  );

  res.status(200).send(file);
};

export default handler;

export const config = {
  api: {
    responseLimit: false,
  },
};
```


```ts
// next.config.mjs

// ...
const config = {
  // ...
  redirects: async () => {
      return [
        {
          source: "/_next/static/chunks/pages/:path*",
          destination: "/api/static?path=:path*",
          permanent: false,
        },
      ];
    }
  // ...
}
// ...
```

With this change, only logged in users are able to fetch `my-private-page-somehash.js`

Not the best solution but easy to implement and gets the job done.
